doc/design/proposal-pre-import.md

```diff
@@ -0,0 +1,97 @@
+# preImport Proposal
+
+## Problem Statement
+
+With the core resolver now synchronous only, any asynchronous resolution work
+is no longer possible as the resolve function must return an immediate
+resolution.
+
+For loaders that asynchronously obtain resolution information, it would be
+useful to have a new hook to enable these use cases.
+
+## preImport Hook
+
+The hook has the signature:
+
+```ts
+export async function preImport(specifier: string, context: {
+  conditions: string[],
+  topLevel: boolean,
+  parentURL: string | undefined
+});
+```
+
+The `preImport` hook allows for tracking and asynchronous setup work for every
+top-level import operation. It has no return value, although it can return
```

Comment:
This “top-level” qualifier is worth calling out. What’s the precise definition? Any static or dynamic `import` in the application entry point file? Only dynamic imports in that entry point?

Is that whatever data gets created by `preImport` then available for _all_ `resolve` operations, not just top-level specifiers?

Since this is something of an initialization-of-the-module-graph thing, not run for _every_ import, would it make sense for this hook to just run once and be passed in an array of specifiers? Like `preImport(specifiers: string[], context: object)` where `specifiers` would be the list of all top-level import specifiers.

Reply:
The definition is any import which calls [ECMA function 16.2.1.5.2 Evaluate](https://tc39.es/ecma262/#sec-moduleevaluation).

It is called for all dynamic imports and the main entry point, and any module worker thread instantiation, and will be called with the exact specifier passed to those.

It's singular and not a list because the timing is unique to each specifier.

Reply:
[ECMA function 16.2.1.5.2 Evaluate](https://tc39.es/ecma262/#sec-moduleevaluation) is incomprehensible to me.

Say an application consists of these two files. Which imports get `preImport` called?

```js
// entry.js
import 'a';
await import('b');
import './c.js';

// c.js
import 'd';
await import('e');
```

Reply:
In the example, it would get called for `entry.js`, then `'b'` when that import is hit, and then `'e'` when that dynamic import is hit.

Reply:
Okay, so perhaps the documentation should say something along the lines of this:

> The `preImport` hook is called for the root of every tree of modules imported by a dynamic `import`: the initial entry point itself (the file passed to `node` on the command line) and any _dynamic_ `import()` calls within. The way ES modules work is that an entry point is read and statically analyzed, and all the static `import` statements are themselves analyzed and followed and _their_ files read and analyzed, and so on until all the leaves of the tree have been parsed. Then all this code is evaluated; and during this evaluation phase the dynamic `import()` statements are followed, which uncover new trees of modules and sub-modules. The same analysis and evaluation process then begins for each of them. `preImport` happens after the analysis but before the evaluation of each set of modules.

I would think that it could be called with all the specifiers within the tree of modules for which it’s being called? So in this example, the first call could be for `entry.js`, `a`, `c.js` and the second call would be for `b`, `d` and the third call would be for `e`.

Reply:
> In the example, it would get called for `entry.js`, then `'b'` when that import is hit, and then `'e'` when that dynamic import is hit.

why wouldn't it get called for `'a'`, `'./c.js'`, and `'d'`? Standard imports like that can still be async anyway, as I understand it? From @GeoffreyBooth 's comment before, I understood that the _only_ place where it would not be called is when a user invokes `import.meta.resolve` (which I think is a problem, but if this isn't even called when doing a regular import then I'm struggling to see the point)